File: JanperCab/3-Application/Dtos/DuraformMiscComponent/DuraformMiscComponentDto.cs
namespace _3_Application.Dtos.DuraformMiscComponent
{
    public abstract class DuraformMiscComponentDto
    {
        public int Id { get; set; }

        public int DuraformEnquiryId { get; set; }

        public int Quantity { get; set; }

        public decimal UnitPrice { get; set; }

        public decimal SubTotal { get; set; }

        public decimal TotalDiscount { get; set; }

        public decimal TotalPrice { get; set; }
    }
}

File: JanperCab/5-JanperCab/ClientApp/src/app/_models/duraform-misc-component/DuraformMiscComponentDto.ts
import { DuraformEnquiryDto } from 'src/app/_models/enquiry/DuraformEnquiryDto';
import { FormGroup } from '@angular/forms';
import { DuraformMiscTypeEnum } from './../../_enums/DuraformMiscTypeEnum';
import * as _ from 'lodash';

export abstract class DuraformMiscComponentDto {
  $type: string;
  id: number;
  duraformEnquiryId: number;
  quantity: number;
  unitPrice: number;
  subTotal: number;
  totalDiscount: number;
  totalPrice: number;

  constructor() {
    this.$type =
      '_3_Application.Dtos.DuraformMiscComponent.DuraformMiscComponentDto, 3-Application';
  }

  abstract get miscType(): DuraformMiscTypeEnum;
  abstract update(formGroup: FormGroup): void;
  abstract getUnitPrice(duraformEnquiry: DuraformEnquiryDto): number;
  abstract calculateUnitPrice(duraformEnquiry: DuraformEnquiryDto): void;
  abstract toString(): string;

  calculateTotal(duraformEnquiry: DuraformEnquiryDto): void {
    this.subTotal = this.unitPrice * this.quantity;

    this.totalDiscount = _.round(
      (this.subTotal * duraformEnquiry.discountRate) / 100,
      2
    );

    this.totalPrice = this.subTotal - this.totalDiscount;
  }
}
